### ISSUE TITLE
hello, example 06: sound adpcma ,how to play a 3minutes song?

### ISSUE THREAD

**Issue by ng0757** [dciabrin/ngdevkit-examples#8]
Hello, I'm a beginner programmer,
In this example 06 sound adpcma, I press the C key to play lefthook.mp3→adpcma,
I want to play a complete song. I have replaced lefthook.mp3 with a 3minutes song,
What specific changes do I need to make next to achieve this goal?
Can you give a detailed modification method? I don't know which files need to be modified. Like I said I'm a beginner , just for interest.

**Comment by dciabrin** [comment]
So some quick math... if my understanding of the awesome neogeodev wiki is correct [1], ADPCM-A playback is 18.5185Khz, let's assume it's roughly 18000 samples per seconds..
In the sound ROM, ADPCM-A encodes a sample as 4bits. That means each byte of the ROM holds 2 audio samples, and thus 1 second worth of ADPCM-A sound takes 9000bytes in the ROM.

3 minutes being 180s, you'd need 180*9000 = 1620000 bytes or roughly 1.58 MiB.
However the YM2610 only allow playback of ADPCM-A data up to 1MiB long, which would be a bit less than 2 minutes.

Now, if you want to play with that, have a look at 06-sound-adpcma/user_commands.s, for example snd_command_04_action_config:

```
snd_command_04_action_config:
        .dw     0x0                     ; sample start addr >> 8
        .dw     0x2                     ; sample stop addr >> 8
        .db     1                       ; channel 3
        .db     0xdf                    ; l/r output + volume
```

This tells the ym2610 to play a ADPCM-A sample from ROM address 0*256, up to ROM address 0x2*256. You can't play or stop a stop on a position that's not a multiple of 256, hence the weird `>>8` encoding above.

If you want to play a longer sample, you'd have to replace .dw 0x2 with the sample's end address in the ROM.


[1] https://wiki.neogeodev.org/index.php?title=ADPCM

**Comment by ng0757** [comment]
Thank you for your detailed answer.I successfully played a 40sec music.
But there are intermittent noises in it, which makes very uncomfortable.
I tested the lefthook.wav generated by the transformation, and there was no noise there.
Therefore, the noise should appear in the step of wav to ADPCA. 

I checked the Makefile,

samples/lefthook.wav: sfx/lefthook.mp3 | samples
	$(SOX) -V1 $< -c 1 -r 18500 $@ silence 1 0.1 1% reverse silence 1 0.2 1% reverse

How can I change the parameters to make the song convert "perfectly"?

**Comment by dciabrin** [comment]
> samples/lefthook.wav: sfx/lefthook.mp3 | samples
> $(SOX) -V1 $< -c 1 -r 18500 $@ silence 1 0.1 1% reverse silence 1 0.2 1% reverse

Oh this one was just an artifact of the original file, I didn't want to ship a modified source asset, so I had to cut the beginning and the end at compile time.

> 
> How can I change the parameters to make the song convert "perfectly"?

It should be safe to convert your input without any sox filter:

```
$(SOX) -V1 $< -c 1 -r 18500 $@
```

If there's still noise during the ym2610 playback and the wav file doesn't contain it, it might be an adpcm encoding bug. I tried to read carefully the ADPCM encoding algorithm, but we never know.

**Comment by ng0757** [comment]
It's probably an adpcm bug, the wave file is good.
I changed a song and there was no noise after the conversion.
Thanks again~

**Comment by ng0757** [comment]
hello,There are still some situations,I can only play the music for about 56 sec,I did the calculations: 524288/ 18518.5/2=56
So I tried to change it in the makefile
$(VROM): $(SAMPLES:%=samples/%.adpcma) | rom
	cat $(SAMPLES:%=samples/%.adpcma) > $@ && $(TRUNCATE) -s 1048576 $@

make v-rom from 524288 to 1048576  , and here
        .dw     0x0                     ; sample start addr >> 8
        .dw     0x900                     ; sample stop addr >> 8

But the result is still only 56 sec,. Part of the sound beyond 56 seconds is all noise.
if I changed it to 0x799 ,then just play 56sec with no noise.
I guess I didn't put more than 56 seconds into vrom correctly.
what should I do?

**Comment by dciabrin** [comment]
Ok I think this is due to the fact that the example ROM is built to mimic the original puzzledp, which only has a 512KB sound ROM...

This is an interesting side effect of building custom ROMs for emulators: each official game has a slightly different ROM size/configuration, so GnGeo hardcodes all the configuration for known games (mame does that as well).

You can't really tell a NeoGeo emulator to load a custom ROM because there's no generic mechanism to tell it to parse how many ROMs it contains, how big they are, and how to map them to the CPU or GPU address space.. So in ngdevkit-examples I just arbitrarily picked puzzledp, and I build all the example ROMs to mimic puzzledp's ROM configuration in GnGeo: 2 x 1MB character ROM, 1 512KB program ROM, 1 512KB ADPCM sample ROM...

So in order to output a 1MB consecutive sample, you'd need to rename puzzledp.zip to the name of an official game that contains a 1MB ROM. From mamedev's git repo [1], you can see that quiadais would work.

Bottom line: if you want to play a 1MB-long ADPCM-A sample, you'd need to create a ROM named quizdais.zip, which contain files named:
  023-p1.bin
  023-s1.bin
  023-m1.bin
  023-v1.bin
  023-c1.bin
  023-c2.bin

[1] https://github.com/mamedev/mame/blob/master/hash/neogeo.xml

**Comment by ng0757** [comment]
According to your instructions, I successfully played a perfect 60sec song with quizdais.zip.
I checked that the maximum vrom capacity record seems to be 16MB. Can it exceed? such as 32MB, or infinity?

I tried to modify gngeo-nightly/gngeo.dat/rom/puzzledp.drv to increase puzzledp's vrom capacity. First, I tried to change it to 1MB for test,
At address 0xD0 to 0xD3, the content is 00 00 08 00.
I think this should be the capacity of vrom. So I changed it to 00 00 10 00,
Then recompress it back to gngeo.dat and recompile gngeo,
But it didn't succeed in the end, still 56sec when load this puzzledp.zip.
Is my method of expanding puzzledp'vrom imperfect or totally wrong?

**Comment by ng0757** [comment]
I also modified the romrc file
#mame set of  puzzledp 
game  puzzledp  MVS "Puzzle De Pon" 
202-v1.bin 0x0 0x100000 NORM

**Comment by dciabrin** [comment]
> According to your instructions, I successfully played a perfect 60sec song with quizdais.zip.
> I checked that the maximum vrom capacity record seems to be 16MB. Can it exceed? such as 32MB, or infinity?
> 

If I'm interpreting the neogeodev wiki [1] correctly, it seems the ym2610 on the neogeo is hooked to a ROM bank/demultiplexer chip (neo-pcm), which allow adressing 4 different VROM of 4MB max? (only 22bits of adress space). I believe that's where the 16MB comes from.

Practically the emulator abstracts the demultiplexing away, but the limit is still there.

> I tried to modify gngeo-nightly/gngeo.dat/rom/puzzledp.drv to increase puzzledp's vrom capacity. First, I tried to change it to 1MB for test,
> At address 0xD0 to 0xD3, the content is 00 00 08 00.
> I think this should be the capacity of vrom. So I changed it to 00 00 10 00,
> Then recompress it back to gngeo.dat and recompile gngeo,
> But it didn't succeed in the end, still 56sec when load this puzzledp.zip.
> Is my method of expanding puzzledp'vrom imperfect or totally wrong?

Make sure you also tweak bites at address 0x1dc to 0x1af, it's the size of the 202-v1.bin which is loaded in what gngeo calls "region 3". You modified the space for region 3 but not the size of the rom file to load.

**Comment by ng0757** [comment]
I only changed the total capacity of all vrom and forgot to modify the capacity of V1.
Ah! this is a stupid mistake.
I finally changed it to 8M capacity, including 5 songs.

By the way, only the 1st song had continuous noise,maybe the song itself has a problem.

and thanks again.

**Comment by dciabrin** [comment]
Ok seems this issue is resolved, closing accordingly.
